README.md
# formista

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]


This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what
PSRs you support to avoid any confusion with users and contributors.

## Install

Via Composer

``` bash
$ composer require evista/formista
```

## Usage

`<h3>Simple form</h3><div>To create a form, simply add a new class with any name that extends BaseForm class:</div>
 ```php
 use Evista\Formista\ValueObject\FormField;
 use Evista\Formista\Form\BaseForm;
 
 
 class ExampleForm extends BaseForm
 {
   //...
 }
 ```
 This class is not very useful without form elements. To add any field, for example a hidden input you have to implement a class method, called&nbsp;generateFields().&nbsp;
 ```php
  public function generateFields(){
         // Name field
         $name = new FormField(FormField::TYPE_TEXT_INPUT);
         $name
             ->setName('name')
             ->setAttributes(['placeholder' => 'Minta János', 'id' => 'name']);
       
         $this->formFields['name'] = $name;
     }
 ```
 Now your ExampleForm has an input field, called name. If you want do display the form, just use this class:
 ```php
 $form = new ExampleForm();
 
 print '<input type="'.$form->getField('name')->getType().'" name="'.$form->getField('name')->getName().'" value="'.$form->getField('name')->getValue().'"/>';
 
 
 ```
 <div>After the user posts this to the server, you can rebuild the form the same way as before:</div>
 ```php
 $form = new ExampleForm();
 ```
 But this time every field value will be populated with what the user sent.
 ```php
 $form->getField('name')->getValue() // Bálint 
 ```
 It’s not just more convenient but makes your code more readable and conceivable. But Fom API can do even more: you can set up validations;
 ```php
 // Phone
 $phone = new FormField(FormField::TYPE_TEXT_INPUT);
 $phone
     ->setName('phone')
     ->setAttributes(['placeholder' => '+36 30 111 2222', 'id' => 'phone'])
     ->setSanitizationCallback(function($value){
         // only numbers, whitespaces and +
         return trim(preg_replace($this->phoneNumberPattern, '', $value));
     })
     ->setValidationCallback(
         function($value){
             // Length constrain
             if(strlen($value)<5){
                 return 'Telephone number is not valid';
             }
             // Regex constrain
             if(preg_match($this->phoneNumberPattern, $value)){
                 return 'Telephone number is not valid';
             }
             // False means it's OK!
             return false;
         }
     )
     ->setMandatory(true);
 $this->formFields['phone'] = $phone;
 ```
 With setValidationCallback() method you can define a function that gets the submitted value as $value. Returning false means that there’s no problem with the submitted data. The form will run all validations on all input when you call the validate() method.&nbsp;
 ```php
 $form->validate();
 ```
 Any field can be set mandatory with the setMandatory() method.<div>Validation is not automatic so don’t forget to call the validate() method.</div>
 ```php
 $form = new ExampleForm();
 if(count($errors = $form->validate()) > 0){
   //something is wrong, the messages are in the $errors array
 }
 // It's valid yeah
 ```
 You can set santitization callback which will be called autmoatically when you instantiate a form.
 Csrf protection is fully automatic also so you don’t need to do anything.<div><br></div>
 <h3>Ajax forms</h3><div><br></div>
 
 <br>
 
 
 
 
 
 
 
 
 
 
 
 
 


## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email sera.balint@e-vista.hu instead of using the issue tracker.

## Credits

- [Balint Sera][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/league/formista.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/thephpleague/formista/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/thephpleague/formista.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/thephpleague/formista.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/league/formista.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/league/formista
[link-travis]: https://travis-ci.org/thephpleague/formista
[link-scrutinizer]: https://scrutinizer-ci.com/g/thephpleague/formista/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/thephpleague/formista
[link-downloads]: https://packagist.org/packages/evista/formista
[link-author]: https://github.com/:author_username
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `formista` will be documented in this file

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/BaseFormTest.php
<?php

namespace Evista\Formista\Test;

use Evista\Formista\Form\ExampleForm;


/**
 * Created by PhpStorm.
 * User: balint
 * Date: 2015. 10. 15.
 * Time: 12:13
 */
class BaseFormTest extends \PHPUnit_Framework_TestCase
{
    public function testTest(){
        $form = new ExampleForm();
        // Assertion for submit field
        $this->assertInstanceOf('Evista\Formista\ValueObject\FormField', $form->getField('submit'));
        $this->assertEquals('Elküldöm', $form->getField('submit')->getValue());
        $this->assertEquals('submit', $form->getField('submit')->getType());
    }

    public function testCSRF(){
        $form = new ExampleForm();

        $this->assertNotEmpty($form->getField('nonce')->getValue());

    }

    public function testClassName(){
        $form = new ExampleForm();

        $this->assertEquals('ExampleForm', $form->getField('class')->getValue());
    }

}

src/Form/BaseForm.php
<?php

namespace Evista\Formista\Form;

use Evista\Formista\ValueObject\FormField;

/**
 * Created by PhpStorm.
 * User: balint
 * Date: 2015. 10. 14.
 * Time: 9:56
 */
abstract class BaseForm
{
    private $nonceKey = 'djlKJdlkjei877798a7lskdjf';
    private $nonceValue;
    private $submittedData;
    protected $formFields = [];
    protected $templateVars = ['form_fields' =>[]];
    protected $templateName;
    protected $onSubmitCallable;

    /**
     * This is where a child have to define it's fields
     * @return mixed
     */
    abstract function generateFields();


    public function __construct(){
        // Important: add csrf token to every form
        $this->addCSRFTokenField();

        // Set up form class (after submitting we need to know what class to initialize
        $classSelf = new \ReflectionClass($this);
        $className = $classSelf->getShortName();
        $fieldName = 'class';
        $classNameField = new FormField(FormField::TYPE_HIDDEN);
        $classNameField
            ->setName($fieldName)
            ->setValue($className)
            ->setMandatory(true);

        $this->addField($fieldName, $classNameField);


        // Setup submission
        $this->setSubmittedDatasFromPost();

        // Add child fields
        $this->generateFields();

        // Add fields to template variables
        $this->addFieldsToTemplateVars();

        // Populate if we are after submission
        $this->populateFields();

    }


    /**
     * Handles submission - it's better not to call automatically
     * @throws \NoCallbackSetException
     */
    public function handleSubmission(){
        // When posted
        if (null !== $this->getSubmittedData()['nonce']) {
            $this->runOnSubmit();
        }
    }

    public function onSubmit(Callable $callable){
        $this->onSubmitCallable = $callable;
    }

    private function runOnSubmit(){
        if(null === $this->onSubmitCallable){
            throw new \NoCallbackSetException('onSubmit callable is not set');
        }
        $callable = $this->onSubmitCallable;
        $callable();
    }



    /**
     * Get submitted values (independently of submission method eg. ajax / simple)
     */
    private function setSubmittedDatasFromPost(){
        if(null !== $_POST){
            // If ajax, check formData parameter
            if(!empty($_SERVER['HTTP_X_REQUESTED_WITH']) && strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) == 'xmlhttprequest') {
                $keyValuePairs = explode('&', $_POST['formData']);
                array_walk($keyValuePairs, function($value){
                    list($key, $postValue) = explode('=', $value);
                    $this->submittedData[$key] = urldecode($postValue);
                });
            }
            else{
                $this->submittedData = $_POST;
            }
        }
    }

    /**
     * Add CSRF token hidden input field
     */
    private function addCSRFTokenField(){
        $this->nonceValue = $this->createNonce();
        $nonce = new FormField(FormField::TYPE_HIDDEN);
        $nonce
            ->setName('nonce')
            ->setValue($this->createNonce())
            ->setValidationCallback(function($value){
                if(function_exists('wp_verify_nonce')){

                    if(!wp_verify_nonce($value, $this->nonceKey)){
                        throw new \Exception('Unauthorized request');
                    }
                }

                // Use own csrf token
                else{
                    if (!isset($_SESSION['csrf_tokens'][$value])) {
                        throw new \Exception('Unauthorized request');
                    }
                    else{
                        unset($_SESSION['csrf_tokens'][$value]);
                    }
                }

                return false;
            })
            ->setMandatory(true);
        $key = 'nonce';

        $this->addField($key, $nonce);
    }


    /**
     * Create nonce
     * @return string
     */
    private function createNonce(){
        if(function_exists('wp_create_nonce'))
            return wp_create_nonce($this->nonceKey);

        $nonce = md5(microtime(true).$this->nonceKey);
        if (empty($_SESSION['csrf_tokens'])) {
            $_SESSION['csrf_tokens'] = array();
        }
        $_SESSION['csrf_tokens'][$nonce] = true;
        return $nonce;
    }

    /**
     * populates form from POST after submission
     */

    public function populateFields(){
        if(count($this->submittedData)<1) return;
        array_map(function(FormField &$field){
            if(isset($this->submittedData[$field->getName()])){
                $raw = $this->submittedData[$field->getName()];

                $sanitized = $field->sanitize($raw);
                $field->setValue($sanitized);
            }else{
                // Unset value (see: checkboxes where value only sent when checkbox was checked
                if($field->getType() == FormField::TYPE_CHECKBOX){
                    $field->setValue(null);
                }

            }
        },
        $this->getFields());
    }

    /**
     * Validate form input
     * @return mixed
     */
    public function validate(){
        $errors = [];
        array_map(function(FormField $field) use (&$errors){
            if(isset($this->submittedData[$field->getName()])){
                // is it mandatory and empty?
                if($field->isMandatory() && strlen($this->submittedData[$field->getName()]) < 1){
                    $errors[$field->getName()] = [
                        'field' => $field->getName(),
                        'error' => "This is mandatory.",
                    ];

                    // Go to the next field, no need to validate
                    return true;
                }


                $validationResult = $field->validate();
                if($validationResult){
                    $errors[$field->getName()] =
                        [
                            'field' => $field->getName(),
                            'error' => $validationResult
                        ];
                }
                return false;
            }
        },
        $this->getFields());

        // Write to the template vars
        $this->addToTemplateVars($errors, 'form_errors');

        return $errors;
    }

    /**
     * @return mixed
     */
    public function getSubmittedData()
    {
        return $this->submittedData;
    }

    /**
     * Get templateVars
     * @return array
     */
    public function getTemplateVars(){
        return $this->templateVars;
    }

    /**
     * Set templateVars
     * @param array $templateVars
     * @return $this
     */
    public function setTemplateVars(Array $templateVars)
    {
        $this->templateVars = $templateVars;

        return $this;
    }

    /**
     * Add element to template vars
     * @param $element
     * @param null $key
     * @return $this
     */
    public function addToTemplateVars($element, $key = null)
    {
        $this->templateVars[$key] = $element;

        return $this;
    }

    /**
     * Get fields
     * @return array
     */
    public function getFields()
    {
        return $this->formFields;
    }

    /**
     * Add a new field to the form
     * @param $key
     * @param $field
     * @return $this
     */
    public function addField($key, $field)
    {
        $this->formFields[$key] = $field;

        return $this;
    }

    /**
     * Get a field
     * @param $key
     * @return mixed
     */
    public function getField($key){
        return $this->formFields[$key];
    }

    /**
     * Adds fields to template vars
     */
    private function addFieldsToTemplateVars(){
        $this->templateVars['form_fields'] = array_merge($this->formFields, $this->templateVars['form_fields']) ;
    }

    /**
     * @return mixed
     */
    public function getTemplate()
    {
        return $this->templateName;
    }

    /**
     * @param mixed $template
     * @return BaseForm
     */
    public function setTemplate($template)
    {
        $this->templateName = $template;

        return $this;
    }


}

src/ValueObject/FormField.php
<?php
/**
 * Created by PhpStorm.
 * User: balint
 * Date: 2015. 10. 14.
 * Time: 10:21
 */

namespace Evista\Formista\ValueObject;


class FormField
{
    const TYPE_TEXT_INPUT = 'input';
    const TYPE_SUBMIT = 'submit' ;
    const TYPE_HIDDEN = 'hidden';
    const TYPE_CHECKBOX = 'checkbox';
    const TYPE_TEXTAREA = 'textarea';

    private $type;
    private $attributes = [];
    private $name;
    private $value;
    private $default;
    private $tagName = 'input';
    private $mandatory = FALSE;
    private $sanitizationCallback;
    private $validationCallback;
    private $label; // only checkboxes self::TYPE_CHECKBOX

    public function __construct($type){
        $this->type = $type;

        // Default callbacks do nothing, just returns the original
        $this->sanitizationCallback = function($value){ return $value; };
        $this->validationCallback = function($value){ return false; };

        switch($this->type){
            case self::TYPE_TEXTAREA:
                $this->tagName = 'textarea';
                break;
        }
    }

    /**
     * @return mixed
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * @param mixed $type
     * @return FormField
     */
    public function setType($type)
    {
        $this->type = $type;

        return $this;
    }

    /**
     * @return array
     */
    public function getAttributes()
    {
        return $this->attributes;
    }

    /**
     * @param array $attributes
     * @return FormField
     */
    public function setAttributes($attributes)
    {
        $this->attributes = $attributes;

        return $this;
    }

    /**
     * @return mixed
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param mixed $name
     * @return FormField
     */
    public function setName($name)
    {
        $this->name = $name;

        return $this;
    }

    /**
     * @return mixed
     */
    public function getValue()
    {
        return $this->value;
    }

    /**
     * @param mixed $value
     * @return FormField
     */
    public function setValue($value)
    {
        $this->value = $value;

        return $this;
    }

    /**
     * @return mixed
     */
    public function getDefault()
    {
        return $this->default;
    }

    /**
     * @param mixed $default
     * @return FormField
     */
    public function setDefault($default)
    {
        $this->default = $default;

        return $this;
    }

    /**
     * @return boolean
     */
    public function isMandatory()
    {
        return $this->mandatory;
    }

    /**
     * @param boolean $mandatory
     * @return FormField
     */
    public function setMandatory($mandatory)
    {
        $this->mandatory = $mandatory;

        return $this;
    }

    /**
     * @param $inputValue
     * @return closure's return value
     */
    public function getSanitizationCallback($inputValue)
    {
        return $this->sanitizationCallback;
    }

    /**
     * @param \Closure $sanitizationCallback
     * @return FormField
     */
    public function setSanitizationCallback(\Closure $sanitizationCallback)
    {
        $this->sanitizationCallback = $sanitizationCallback;

        return $this;
    }

    /**
     * @return mixed
     */
    public function getLabel()
    {
        return $this->label;
    }

    /**
     * @param mixed $label
     * @return FormField
     */
    public function setLabel($label)
    {
        $this->label = $label;

        return $this;
    }

    /**
     * @return string
     */
    public function getTagName()
    {
        return $this->tagName;
    }

    /**
     * @param string $tagName
     * @return FormField
     */
    public function setTagName($tagName)
    {
        $this->tagName = $tagName;

        return $this;
    }




    public function sanitize($inputValue){
        $function = $this->sanitizationCallback;
        return $function($inputValue);
    }

    /**
     * @return mixed
     */
    public function getValidationCallback()
    {
        return $this->validationCallback;
    }

    /**
     * @param mixed $validationCallback
     * @return FormField
     */
    public function setValidationCallback($validationCallback)
    {
        $this->validationCallback = $validationCallback;

        return $this;
    }


    public function validate(){
        $validateFunction = $this->validationCallback;
        return $validateFunction($this->getValue());
    }



}

src/Form/ExampleForm.php
<?php
/**
 * Created by PhpStorm.
 * User: balint
 * Date: 2015. 10. 14.
 * Time: 10:01
 */

namespace Evista\Formista\Form;

use Evista\Formista\ValueObject\FormField;
use Evista\Formista\Form\BaseForm;


class ExampleForm extends BaseForm
{
    private $phoneNumberPattern = '/[^0-9,\+\s]/';

    /**
     * This is where form elements are defined
     */
    public function generateFields(){
        // Name field
        $name = new FormField(FormField::TYPE_TEXT_INPUT);
        $name
            ->setName('name')
            ->setAttributes(['placeholder' => 'Minta János', 'id' => 'name'])
            ->setMandatory(true);
        $this->formFields['name'] = $name;

        // Phone
        $phone = new FormField(FormField::TYPE_TEXT_INPUT);
        $phone
            ->setName('phone')
            ->setAttributes(['placeholder' => '+36 30 111 2222', 'id' => 'phone'])
            ->setSanitizationCallback(function($value){
                // only numbers, whitespaces and +
                return trim(preg_replace($this->phoneNumberPattern, '', $value));
            })
            ->setValidationCallback(
                function($value){
                    // Length constrain
                    if(strlen($value)<5){
                        return 'Telephone number is not valid';
                    }

                    // Regex constrain
                    if(preg_match($this->phoneNumberPattern, $value)){
                        return 'Telephone number is not valid';
                    }

                    // False means it's OK!
                    return false;
                }
            )
            ->setMandatory(true);
        $this->formFields['phone'] = $phone;

        // Email
        $email = new FormField(FormField::TYPE_TEXT_INPUT);
        $email
            ->setName('email')
            ->setAttributes(['placeholder' => 'minta.janos@info.hu', 'id' => 'email'])
            ->setMandatory(true)
            ->setSanitizationCallback(function($value){return sanitize_email($value);})
            ->setValidationCallback(
                function($value){
                    //var_dump('validating email: '.$value);
                    if(!is_email($value)){
                        return 'Email is not valid';
                    }

                    // False means it's OK!
                    return false;
                }
            );
        $this->formFields['email'] = $email;

        // Submit
        $submit = new FormField(FormField::TYPE_SUBMIT);
        $submit->setValue('Elküldöm');
        $this->formFields['submit'] = $submit;

    }

}